Create broken-linkcheck.yml

ggold7046 opened the pull request:
Created a Github Action to run the Sphinx linkcheck monthly. 

- [x] closes #xxxx (Replace xxxx with the GitHub issue number) #45409
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

ggold7046 committed:
Create broken-linkcheck.yml

Created a Github Action to run the Sphinx linkcheck monthly.
#45409

mroeschke commented:
Could you make this workflow run in this PR to check the output?

ggold7046 commented:
> Could you make this workflow run in this PR to check the output?

Could you tell me how to do that ? Do I need to change the yaml code ?

mroeschke commented:
Yes see https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#on or other workflows in this repo

ggold7046 committed:
Update broken-linkcheck.yml

ggold7046 committed:
Update broken-linkcheck.yml

ggold7046 commented:
I don't see any `run workflow` button. [manually-running-a-workflow](https://docs.github.com/en/actions/using-workflows/manually-running-a-workflow)

ggold7046 committed:
Update broken-linkcheck.yml

ggold7046 committed:
Update broken-linkcheck.yml

ggold7046 committed:
Update broken-linkcheck.yml

ggold7046 commented:
@pmeier could you see this please ?

pmeier commented on .github/workflows/broken-linkcheck.yml:
```diff
@@ -0,0 +1,30 @@
+    workflow_dispatch: # Enable manual triggering
+    #schedule:
+    # Run monthly on the 1st day of the month
+        #- cron: '0 0 1 * *'
```
What @mroeschke advocates here is to also have a trigger for this PR to run this. You can do that by

```suggestion
        #- cron: '0 0 1 * *'
    pull_request:
```

With that, the check would be run on every PR. This is not what we want. Instead we should only run it if the PR touches a file that is directly linked to the test that we are running. My suggestion here is

```suggestion
        #- cron: '0 0 1 * *'
    pull_request:
      paths:
        - ".github/workflows/broken-linkcheck.yml"
        - "doc/make.py"  
```

This way we always run the check if someone touches the workflow file or modifies the make script and thus potentially touches the linkcheck functionality.

ggold7046 replied:
I have a doubt. Why we are using `"- doc/make.py" ` in the `paths:` ?

pmeier replied:
Because if someone changes and accidentally breaks the linkcheck command in `doc/make.py` the CI would not flag it in the PR. I'll leave it up to the maintainers to decide if they want this or not.

ggold7046 replied:
> Because if someone changes and accidentally breaks the linkcheck command in `doc/make.py` the CI would not flag it in the PR.

Didn't understood it properly. Could you explain this little more ?

pmeier commented on .github/workflows/broken-linkcheck.yml:
```diff
@@ -0,0 +1,30 @@
+---
```
This is not needed
```suggestion
```

pmeier commented on .github/workflows/broken-linkcheck.yml:
```diff
@@ -0,0 +1,30 @@
+---
+name: Broken linkcheck
```
Name sounds a little weird to me. Why not just

```suggestion
name: Linkcheck
```

pmeier commented on .github/workflows/broken-linkcheck.yml:
```diff
@@ -0,0 +1,30 @@
+        - name: Install requirements-dev.txt
+          run: pip install -r requirements-dev.txt
+
+        - name: Run linkcheck script
+          run: python make.py linkcheck
```
This won't work since the current directory is the project root. So it should either be

```suggestion
        - name: Run linkcheck script
          run: python doc/make.py linkcheck
```

or even better

```suggestion
        - name: Run linkcheck script
          working-directory: ./doc
          run: python make.py linkcheck
```

ggold7046 committed:
Update .github/workflows/broken-linkcheck.yml

Co-authored-by: Philip Meier <github.pmeier@posteo.de>

ggold7046 committed:
Update .github/workflows/broken-linkcheck.yml

Co-authored-by: Philip Meier <github.pmeier@posteo.de>

ggold7046 committed:
Update .github/workflows/broken-linkcheck.yml

Co-authored-by: Philip Meier <github.pmeier@posteo.de>

ggold7046 committed:
Update .github/workflows/broken-linkcheck.yml

Co-authored-by: Philip Meier <github.pmeier@posteo.de>

ggold7046 committed:
Update broken-linkcheck.yml

pmeier commented on .github/workflows/broken-linkcheck.yml:
```diff
@@ -0,0 +1,34 @@
+            cache: 'pip'
+            cache-dependency-path: 'requirements-dev.txt'
+
+        - name: Install requirements-dev.txt
+          run: pip install -r requirements-dev.txt
```
CI is failing, because you are missing the `pandas` installation step. For example

https://github.com/pandas-dev/pandas/blob/d01669f5cdfff4edc889994331901b1f97946fdd/.github/workflows/docbuild-and-upload.yml#L43-L47

ggold7046 committed:
Update broken-linkcheck.yml

ggold7046 committed:
Update broken-linkcheck.yml

pmeier commented on .github/workflows/broken-linkcheck.yml:
```diff
@@ -0,0 +1,40 @@
+        - name: Build Pandas
+          uses: ./.github/actions/build_pandas
+
+        - name: Setup Python
+          id: setup_python
+          uses: actions/setup-python@v4
+          with:
+            python-version: '3.10'
+            cache: 'pip'
+            cache-dependency-path: 'requirements-dev.txt'
+
+        - name: Install requirements-dev.txt
+          run: pip install -r requirements-dev.txt
```
You need to drop these know that we have a conda env and build in there.

```suggestion
```

ggold7046 committed:
Update .github/workflows/broken-linkcheck.yml

Co-authored-by: Philip Meier <github.pmeier@posteo.de>

ggold7046 commented:
Linkcheck failed . It says 

```

File "/home/runner/work/pandas/pandas/doc/make.py", line 23, in <module>
    import docutils
ModuleNotFoundError: No module named 'docutils'

```

pmeier commented:
Sorry, you also need to put 

https://github.com/pandas-dev/pandas/blob/89bd5695df3ccf5b9186b4645118656656831593/.github/workflows/docbuild-and-upload.yml#L33-L37

inside the workflow to automatically activate the conda env.

ggold7046 committed:
Update broken-linkcheck.yml

ggold7046 commented:
@pmeier ,  there are many broken links. How to fix them ? 
The linkcheck is showing proper output , then why the ci is showing it as failure ?

pmeier commented on .github/workflows/broken-linkcheck.yml:
```diff
@@ -0,0 +1,33 @@
+
+        - name: Build Pandas
+          uses: ./.github/actions/build_pandas
+
```
Nit: remove extra blank line

```suggestion
```

pmeier commented on .github/workflows/broken-linkcheck.yml:
```diff
@@ -0,0 +1,33 @@
+
+        - name: Run linkcheck script
+          working-directory: ./doc
+          run: python make.py linkcheck
```
Instead of dumping everything to the console, we could just forward the broken links

```suggestion
          run: python make.py linkcheck | grep broken
```

Alternatively, we could also store the output in a file and show the broken ones on failure:

```suggestion
          run: python make.py linkcheck | tee linkcheck.txt
          
        - name: Display broken links
          if: failure()
          working-directory: ./doc
          run: grep broken linkcheck.txt
```

ggold7046 replied:
There are also lot of anchor error and client errors.

ggold7046 replied:
> failure()

Don't we need ${{}} for failure() syntax ?

pmeier replied:
> Don't we need ${{}} for failure() syntax ?

No, not inside [`if`](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepsif)

> When you use expressions in an `if` conditional, you may omit the `${{ }}` expression syntax because GitHub Actions automatically evaluates the `if` conditional as an expression. However, this rule does not apply everywhere.

pmeier reviewed (commented):
> there are many broken links. How to fix them ? 

Well, find the source that contains the broken link, and replace it with a working one.

> The linkcheck is showing proper output , then why the ci is showing it as failure ?

The whole point of the linkcheck CI is to fail in case there are broken links. Meaning, this PR cannot be merged until the links are fixed.

Alternatively, `sphinx-build linkcheck` has [options to ignore certain failures](https://www.sphinx-doc.org/en/master/usage/configuration.html#options-for-the-linkcheck-builder). You could also add ignores for all broken links for now and chip away at the ignore list later on. The options need to go into [`conf.py`](https://github.com/pandas-dev/pandas/blob/main/doc/source/conf.py).

If I were you, I would probably write a small script to parse the output of the CI run, which you can download [here](https://github.com/pandas-dev/pandas/commit/8484c630aa685b7d03094dca0834a1785a910e11/checks/17143503456/logs), to automatically create the ignore list.

ggold7046 committed:
Update .github/workflows/broken-linkcheck.yml

Co-authored-by: Philip Meier <github.pmeier@posteo.de>

ggold7046 committed:
Update .github/workflows/broken-linkcheck.yml

Co-authored-by: Philip Meier <github.pmeier@posteo.de>

ggold7046 committed:
Update broken-linkcheck.yml

ggold7046 committed:
Update broken-linkcheck.yml

ggold7046 commented:
> You could also add ignores for all broken links for now and chip away at the ignore list later on.

But we have to create a new issue (and delete the old one) reporting all the errors and label and Docs and Good First Issue (no. 3)
 https://github.com/pandas-dev/pandas/issues/45409#issue-1105272826

ggold7046 commented:
@pmeier , could you tell me how to write the script for no 3 problem (new issue (and delete the old one) Docs and Good First Issue etc etc) ?
 And what about this https://github.com/pandas-dev/pandas/pull/55246#discussion_r1339864162 ? The anchor issues ?

pmeier commented on .github/workflows/broken-linkcheck.yml:
```diff
@@ -0,0 +1,37 @@
+
+        - name: Run linkcheck script
+          working-directory: ./doc
+          run: python make.py linkcheck | tee linkcheck.txt
```
My suggestion actually broke the check. Since we are piping to `tee`, the non-zero exit code, i.e. the failure of the linkcheck, is lost.

```suggestion
          shell: bash
          run: |
            set -o pipefail
            python make.py linkcheck | tee linkcheck.txt
```

ggold7046 replied:
Could you please explain this change in these `32-35` lines ?

pmeier reviewed (commented):
> how to write the script for no 3 problem (new issue (and delete the old one) Docs and Good First Issue etc etc) ?

There is no point on starting with this until this PR is merged.

> And what about this https://github.com/pandas-dev/pandas/pull/55246#discussion_r1339864162 ? The anchor issues ?

All errors that are surfaced by the new CI check need to be either fixed or ignored before this PR can be merged. As suggested previously, I would ignore everything for now and open an issue to get some help in fixing the links. The ignore list can likely be build programmatically from the output of the CI job.

ggold7046 committed:
Update .github/workflows/broken-linkcheck.yml

Co-authored-by: Philip Meier <github.pmeier@posteo.de>

ggold7046 commented:
> The ignore list can likely be build programmatically from the output of the CI job.

Do I need to put that script into [conf.py](https://github.com/pandas-dev/pandas/blob/main/doc/source/conf.py) ?

ggold7046 committed:
Update broken-linkcheck.yml

ggold7046 commented:
@pmeier

pmeier commented on .github/workflows/broken-linkcheck.yml:
```diff
@@ -0,0 +1,41 @@
+
+        - name: Run linkcheck script
+          working-directory: ./doc
+          shell: bash
```
Sorry, I broke this in my last suggestion. We need to use the default shell here to get access to the conda env.

```suggestion
```

pmeier reviewed (commented):
> Do I need to put that script into [conf.py](https://github.com/pandas-dev/pandas/blob/main/doc/source/conf.py) too? 

No. The script is a one-time usage. When the ignore list is generated, we no longer need it. You can post it in this thread if its not too long or other throw it into a Github Gist for future reference.

> where shoould I put that into the conf.py file ?

What is "that" here? The ignore list? I would just put it at the end of `conf.py`. If someone complains later, we can always move it.

ggold7046 committed:
Update .github/workflows/broken-linkcheck.yml

Co-authored-by: Philip Meier <github.pmeier@posteo.de>

ggold7046 commented:
@pmeier , I have written the ignored list script in the drop-down. Should I create another PR for it or should I attach it in here?

<Details>

```
import re
import requests

# Fetch the content of the ci file
pipeline_url = ("https://pipelinesghubeus22.actions.githubusercontent.com/"
                "xZyE9jtmkxWlfCAbyu1SHPJOlsa2huNFYcxohSTomy6EbdNZT9/"
                "_apis/pipelines/1/runs/531865/signedlogcontent/2?"
                "urlExpires=2023-10-08T10%3A02%3A18.0563894Z&"
                "urlSigningMethod=HMACV1&"
                "urlSignature=9icPkVbCE2Ya0M5%2FY03N8fkFuDfYBn%2F"
                "DyJ93o2R4%2BWk%3D")

response = requests.get(pipeline_url)
ignore_patterns = response.text.splitlines()

# List to store broken links
broken_links = []

# Ignore links based on patterns from ci file
for pattern in ignore_patterns:
    # Compile the regular expression pattern
    regex_pattern = re.compile(pattern)

    # Check if any broken link matches the pattern
    ignored_links = [link for link in broken_links if regex_pattern.search(link)]


    print(f"Ignored links matching pattern {pattern}:")
    print(ignored_links)
```

</Details>

pmeier commented:
@ggold7046 As is the script is not executable, since the link is expired. In any case, this script is just for you to automatically build the ignore list that you'll need to put into `conf.py`. Looking at the CI output, we only have roughly 30 broken links. With that number it is still feasible to do it manually.

ggold7046 commented:
I can not see the linkcheck.txt file in the ./doc directory.
> Looking at the CI output, we only have roughly 30 broken links. With that number it is still feasible to do it manually.

I can only see a single 404 from the linkcheck output. Maybe other 404 are fixed.

pmeier commented:
> I can not see the linkcheck.txt file in the ./doc directory.

This file only exists in the CI runner and won't be uploaded or the like.

> I can only see a single 404 from the linkcheck output.

You can see all failures marked as "broken" here: https://github.com/pandas-dev/pandas/actions/runs/6420191351/job/17431740319?pr=55246#step:6:9

ggold7046 commented:
@pmeier , I have opened a new issue for this broken links. Let's take the first issue.

`(whatsnew/v0.13.1: line   22) broken    http://scatterci.github.io/pydata/pandas - 404 Client Error: Not Found for url: http://scatterci.github.io/pydata/pandas`

Should I ask someone to provide me with the correct link and put that correct link into the doc ? How would I find out which page of the doc contains this erroneous link ?

ggold7046 committed:
Update conf.py

Added Ignore list for broken link checks

pmeier commented on doc/source/conf.py:
```diff
@@ -798,3 +798,35 @@ def setup(app):
+
+#### Ignore list for broken link checks
+
+import re
```
This is not an ignore list. Have a look at https://github.com/pandas-dev/pandas/pull/55246#pullrequestreview-1647254547 and follow the instructions.

ggold7046 replied:
I think I didn't understand the proper way of doing the ignore list. Should I take `linkcheck_ignore = [r'http://localhost:\d+/']` and put all the [broken](https://github.com/pandas-dev/pandas/actions/runs/6420191351/job/17431740319?pr=55246#step:6:9) links from the yaml file and put that list inside the conf.py file at the end ? 

What is the use of this [file](https://pipelinesghubeus22.actions.githubusercontent.com/xZyE9jtmkxWlfCAbyu1SHPJOlsa2huNFYcxohSTomy6EbdNZT9/_apis/pipelines/1/runs/531865/signedlogcontent/2?urlExpires=2023-10-10T19%3A08%3A26.1254826Z&urlSigningMethod=HMACV1&urlSignature=Zgha0BunA3OJneuIap0FFToSo%2FK3YhduNbCHSBXb%2Fv8%3D) ?

pmeier replied:
> Should I take `linkcheck_ignore = [r'http://localhost:\d+/']` and put all the [broken](https://github.com/pandas-dev/pandas/actions/runs/6420191351/job/17431740319?pr=55246#step:6:9) links from the yaml file and put that code inside the conf.py file at the end ?

Yes. And add a comment on what this ignore list does including a link to this PR.

> What is the use of this [file](https://pipelinesghubeus22.actions.githubusercontent.com/xZyE9jtmkxWlfCAbyu1SHPJOlsa2huNFYcxohSTomy6EbdNZT9/_apis/pipelines/1/runs/531865/signedlogcontent/2?urlExpires=2023-10-10T19%3A08%3A26.1254826Z&urlSigningMethod=HMACV1&urlSignature=Zgha0BunA3OJneuIap0FFToSo%2FK3YhduNbCHSBXb%2Fv8%3D) ?

Not sure where you are getting the links, but again it is expired.

ggold7046 committed:
Update conf.py

#55246
This is an ignore list for broken links found in CI run checks for broken-linkcheck.yml

pmeier commented on doc/source/conf.py:
```diff
@@ -798,3 +798,40 @@ def setup(app):
+
+####### This is an ignore list for broken links found in CI run checks for broken-linkcheck.yml
+
+linkcheck_ignore = [r'http://']
+linkcheck_ignore = [r'http://scatterci.github.io/pydata/pandas:\d+/']
```
1. You can't set this variable multiple times. You'll need to put multiple entries in a list:

```suggestion
linkcheck_ignore = [
    r'http://',
    r'http://scatterci.github.io/pydata/pandas:\d+/'
    ...
]
```

2. Not sure what you want to achieve with the `:\d+/` at the end of every line. With this, `http://scatterci.github.io/pydata/pandas` will not get matched and in turn ignored.

ggold7046 committed:
Update doc/source/conf.py

Co-authored-by: Philip Meier <github.pmeier@posteo.de>

ggold7046 committed:
Update conf.py

pmeier commented:
https://github.com/pandas-dev/pandas/actions/runs/6500122547/job/17654854642?pr=55246#step:5:149

>  ```
>  Exception occurred:
>   File "/home/runner/micromamba/envs/test/lib/python3.10/sre_parse.py", line 843, in _parse
>     raise source.error("missing ), unterminated subpattern",
> ```

And this is why I [suggested](https://github.com/pandas-dev/pandas/pull/55246#discussion_r1354307474) wrapping everything in `re.escape`. Basically what happened is that something in the ignore list included special characters that were picked up by the regex engine. Since we don't want to do regex matching at this point, we should just match the exact version of the link, hence escaping it.

ggold7046 committed:
Update conf.py

ggold7046 committed:
Update conf.py

ggold7046 commented:
@pmeier

pmeier reviewed (commented):
What does this error tell you?

https://github.com/pandas-dev/pandas/actions/runs/6505505089/job/17669250908?pr=55246#step:5:20

> NameError: name 're' is not defined

ggold7046 committed:
Update conf.py

ggold7046 commented:
> NameError: name 're' is not defined

@pmeier , I have imported the `re` module.

```
name: Linkcheck
on:
    workflow_dispatch: 
    # Enable manual triggering
    schedule:
    # Run monthly on the 1st day of the month
        - cron: '0 0 1 * *'
```
The cron tab is showing this error `Runs at 00:00, on day 1 of the month. Actions schedules run at most every 5 minutes using UTC time.` That is why the yaml file ci is also failing.

pmeier commented on doc/source/conf.py:
```diff
@@ -798,3 +799,42 @@ def setup(app):
+####### This is an ignore list for broken links found in CI run checks for broken-linkcheck.yml
+
+linkcheck_ignore = [
+    re.escape(r'http://'),
+    re.escape(r'http://scatterci.github.io/pydata/pandas'),
+    re.escape(r'http://specs.frictionlessdata.io/json-table-schema/'),
+    re.escape(r'https://'),
```
Ignoring the regex engine for now, the matching roughly checks if a given link, e.g. `https://example.com`, matches any of the rules starting from the left. Since you have `'http://'` and `'https://'` in there, every single HTTP link (which should be at least 99% of them) is ignored: https://github.com/pandas-dev/pandas/actions/runs/6506999729/job/17673609837?pr=55246#step:5:147

```suggestion
    re.escape(r'http://scatterci.github.io/pydata/pandas'),
    re.escape(r'http://specs.frictionlessdata.io/json-table-schema/'),
```

ggold7046 replied:
> Since you have `'http://'` and `'https://'` in there, every single HTTP link (which should be at least 99% of them) is ignored: https://github.com/pandas-dev/pandas/actions/runs/6506999729/job/17673609837?pr=55246#step:5:147

Did not understood this properly. Are you saying that all the `http//` and `https//` links which don't have any host supplied; will be ignored automatically ?

pmeier replied:
Ignoring regex for now, the matching here roughly looks like this:

```python
# This is a link that sphinx finds inside the documentation
import re

def is_ignored(link):
    escaped_link = re.escape(link)

    # linkcheck_ignore is the ignore list in conf.
    for ignore_link_pattern in linkcheck_ignore:
        if escaped_link.startswith(ignore_link_pattern):
             return True
    return False

assert is_ignored("https://example.com")
```

If you have `http://` and `https://` in `linkcheck_ignore` every single link that sphinx finds in the documentation that starts with `http://` and `https://`, which should be almost if not all of them, will be ignored. And that is what you see in the CI output. 

But this is not what we want. We only want to ignore the broken ones. Otherwise the whole linkcheck does effectively nothing.

pmeier reviewed (commented):
Side note: since you are currently not writing regular expressions, but rather want to match plain strings, you can drop the "raw string" specifier, i.e. the `r` in `r'...'`.

ggold7046 committed:
Update conf.py

ggold7046 commented:
> he cron tab is showing this error `Runs at 00:00, on day 1 of the month. Actions schedules run at most every 5 minutes using UTC time.` That is why the yaml file ci is also failing.

And how to rectify this ?

pmeier commented:
> And how to rectify this ?

Could you post a (non-expiring) link to where you got this from? I only see a trailing whitespace issue on the CI builds.

ggold7046 committed:
Update conf.py

ggold7046 committed:
Update broken-linkcheck.yml

ggold7046 commented:
> Could you post a (non-expiring) link to where you got this from?

![Screenshot_2023-19](https://github.com/pandas-dev/pandas/assets/98327736/45b61286-156b-4dae-a1db-dc3bd574b0cb)

There is a blue underline in cron. When hovering it shows the error and showing this [Link](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#onschedule)

![cron](https://github.com/pandas-dev/pandas/assets/98327736/b3e6420c-d4bb-425b-b03d-82a3ed839fdf)

ggold7046 commented:
@pmeier, are you available on pandas-dev slack ?

ggold7046 commented:
@pmeier , are you available on pandas-dev slack ?

pmeier commented on doc/source/conf.py:
```diff
@@ -798,3 +799,40 @@ def setup(app):
+
+####### This is an ignore list for broken links found in CI run checks for broken-linkcheck.yml
+
+linkcheck_ignore = [
```
I see now where you got the `http://` and `https://` from: https://github.com/pandas-dev/pandas/actions/runs/6511956106/job/17688629505?pr=55246#step:6:9

https://github.com/pandas-dev/pandas/blob/a83f6aae255d7d88fe26ac12565c26253a95751b/pandas/io/formats/style_render.py#L998-L1001

These two need to be permanently ignored, since there is no broken link here, but rather an incomplete one.

```suggestion
linkcheck_ignore = [
    '^http://$',
    '^https://$',
```

- the `^` at the front means "The link has to start exactly here. No other character is allowed before"
- the `$` at the front means "The link has to stop exactly here. No other character is allowed after"

Have a look at the [documentation of the `re` module](https://docs.python.org/3/library/re.html). I'm also quite fond of https://regex101.com (don't forget to select Python in the left sidebar).

pmeier reviewed (commented):
> are you available on pandas-dev slack ?

I'm not.

---

Re https://github.com/pandas-dev/pandas/pull/55246#issuecomment-1762052223

I'm not sure what your issue here is. I don't know why your editor underlines the cron entry. Looks ok to me and CI is ok with it as well.

ggold7046 committed:
Update doc/source/conf.py

Co-authored-by: Philip Meier <github.pmeier@posteo.de>

ggold7046 committed:
Update conf.py

ggold7046 commented:
@pmeier , is this PR complete apart from the two ci fails ? 

I corrected a line which exceeded more than 88 characters in the conf.py file, but still the pre-coomit.ci is failing.  I couldn't understand any more errors. Neither I understood the circle ci fail.

pmeier commented:
pre-commit.ci autofix

pmeier commented:
> is this PR complete apart from the two ci fails ?

From my side, yes. But now we need a maintainer to review and approve.

> I corrected a line which exceeded more than 88 characters in the conf.py file, but still the pre-coomit.ci is failing.

You need to comment "pre-commit.ci autofix" for the fix to be applied. I tried in https://github.com/pandas-dev/pandas/pull/55246#issuecomment-1768298310, but I don't have permissions. It is basically complaining that the imports in `conf.py` are not ordered correctly.

> Neither I understood the circle ci fail.

This is most likely unrelated to this PR. I'd wager a guess that if you update your branch from the `main` branch, the issue will go away.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ggold7046 committed:
Merge branch 'main' into patch-1

ggold7046 commented:
@mroeschke, could you take a look at this PR and suggest if anything else is needed to be done ?

ggold7046 commented:
@mroeschke , could you take a look at this PR ?

ggold7046 commented:
@pmeier

mroeschke commented on .github/workflows/broken-linkcheck.yml:
```diff
@@ -0,0 +1,40 @@
+name: Linkcheck
+on:
+    workflow_dispatch:
+    # Enable manual triggering
```
```suggestion
```

mroeschke commented on .github/workflows/broken-linkcheck.yml:
```diff
@@ -0,0 +1,40 @@
+        runs-on: ubuntu-latest
+        defaults:
+          run:
+                shell: bash -el {0}
```
```suggestion
            shell: bash -el {0}
```

mroeschke commented on .github/workflows/broken-linkcheck.yml:
```diff
@@ -0,0 +1,40 @@
+        - name: Checkout
+          uses: actions/checkout@v4
+          with:
+                fetch-depth: 0
```
```suggestion
            fetch-depth: 0
```

mroeschke commented on .github/workflows/broken-linkcheck.yml:
```diff
@@ -0,0 +1,40 @@
+
+        - name: Build Pandas
+          uses: ./.github/actions/build_pandas
+
```
```suggestion
```

mroeschke commented on doc/source/conf.py:
```diff
@@ -798,3 +799,71 @@ def setup(app):
     app.add_directive("autosummary", PandasAutosummary)
+
+
+####### This is an ignore list for broken links,
+####### found in CI run checks for broken-linkcheck.yml
```
Can you just use a 1 line comment here?

mroeschke commented on doc/source/conf.py:
```diff
@@ -798,3 +799,71 @@ def setup(app):
+####### This is an ignore list for broken links,
+####### found in CI run checks for broken-linkcheck.yml
+
+linkcheck_ignore = [
```
Could you apply `re.escape` once to all these links in a for loop?

ggold7046 committed:
Update .github/workflows/broken-linkcheck.yml

Co-authored-by: Matthew Roeschke <10647082+mroeschke@users.noreply.github.com>

ggold7046 committed:
Update .github/workflows/broken-linkcheck.yml

Co-authored-by: Matthew Roeschke <10647082+mroeschke@users.noreply.github.com>

ggold7046 committed:
Update .github/workflows/broken-linkcheck.yml

Co-authored-by: Matthew Roeschke <10647082+mroeschke@users.noreply.github.com>

ggold7046 committed:
Update conf.py

ggold7046 committed:
Update .github/workflows/broken-linkcheck.yml

Co-authored-by: Matthew Roeschke <10647082+mroeschke@users.noreply.github.com>

ggold7046 committed:
Update conf.py

ggold7046 commented:
pre-commit.ci autofix

ggold7046 committed:
Merge branch 'main' into patch-1

ggold7046 commented:
pre-commit.ci autofix

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ggold7046 commented:
@mroeschke , I have worked on your suggestions, please let me know if there are other suggestion.

ggold7046 commented:
@pmeier, could you please look at this linkcheck fail ?

pmeier commented:
@ggold7046 your patch resurfaces an issue we had before: https://github.com/pandas-dev/pandas/pull/55246#discussion_r1362686035. Plus there seems to be a new broken link that needs to be ignored as well.

ggold7046 committed:
Update conf.py

ggold7046 commented:
> your patch resurfaces an issue we had before: [#55246 (comment)](https://github.com/pandas-dev/pandas/pull/55246#discussion_r1362686035).

Didn't we toook care of it already ?

pmeier commented:
Well, you "reverted" the patch by applying `re.escape` to everything in the ignore list again.

ggold7046 committed:
Update conf.py

pmeier commented on doc/source/conf.py:
```diff
@@ -840,4 +840,11 @@ def setup(app):
 ]
 
-linkcheck_ignore = [re.escape(link) for link in links]
+# Exclude specific links from the links list
```
What about to just do

```python
linkcheck_ignore = [
    "^http://$", 
    "^https://$",
    *[
        re.escape(link)
        for link in [
            "http://scatterci.github.io/pydata/pandas",
            "http://specs.frictionlessdata.io/json-table-schema/",
            ...
        ]
     ]
]
```

ggold7046 replied:
can't we omit the `*` ?

Could you tell  why my code failed ?

pmeier replied:
> can't we omit the `*` ?

Nope. The `*` before the `[...]` unpacks it. Meaning, we get a flat list of strings. Try printing `[1, 2, [3, 4]]` and `[1, 2, *[3, 4]]` to see the difference.

> Could you tell why my previous code failed ?

The filtering logic is correct, but ultimately what counts is the value of `linkcheck_ignore`. And you excluded `"^http://$"` and `"^https://$"` from it.

ggold7046 committed:
Update conf.py

ggold7046 committed:
Merge branch 'main' into patch-1

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ggold7046 committed:
Merge branch 'main' into patch-1

ggold7046 commented:
@mroeschke , could you take a look at this PR please ?

mroeschke commented on .github/workflows/broken-linkcheck.yml:
```diff
@@ -0,0 +1,37 @@
+          uses: ./.github/actions/setup-conda
+
+        - name: Build Pandas
+          uses: ./.github/actions/build_pandas
+        - name: Run linkcheck script
```
```suggestion
          uses: ./.github/actions/build_pandas

        - name: Run linkcheck script
```

mroeschke commented on .github/workflows/broken-linkcheck.yml:
```diff
@@ -0,0 +1,37 @@
+name: Linkcheck
+on:
+    schedule:
```
Can you ensure this file uses 2 space indentation instead of 4?

mroeschke reviewed (commented):
2 small comments otherwise LGTM

ggold7046 committed:
Update .github/workflows/broken-linkcheck.yml

Co-authored-by: Matthew Roeschke <10647082+mroeschke@users.noreply.github.com>

ggold7046 committed:
Update broken-linkcheck.yml

ggold7046 committed:
Update conf.py

mroeschke commented:
Thanks @ggold7046 and @pmeier for your patience and effort here

ggold7046 commented:
Thanks @mroeschke . 

Thank you @pmeier for the utmost support. 

@pmeier, how to do the [3rd part](https://github.com/pandas-dev/pandas/issues/45409#issue-1105272826) ?

ggold7046 commented:
@pmeier, how to proceed with the [3rd part](https://github.com/pandas-dev/pandas/issues/45409#issue-1105272826) ?

Almost every time the `linkcheck` file is run there is a new broken link found. Is there a way to automatically append that link into the `ignore list` ?
